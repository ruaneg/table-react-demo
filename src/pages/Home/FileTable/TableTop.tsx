import styled from 'styled-components';
import DownloadButton from '~/components/Buttons/DownloadButton';
import { FileData } from './fileData';

export const StyledTableTop = styled.div`
  padding: 16px;
  border: 1px solid #f1efef;
  border-bottom: none;
  font-size: 1rem;
  display: flex;
`;

const SelectedText = styled.div`
  width: 8rem;
  margin: 0 1rem;
`;

export const TableTop = ({
  selectAll,
  indeterminate,
  selectedRows,
  handleSelectAllChange,
}: {
  selectAll: boolean;
  indeterminate: boolean;
  selectedRows: FileData[];
  handleSelectAllChange: () => void;
}) => {
  const handleDownloadSelected = () => {
    const selectedPaths = selectedRows
      .map((file) => `${file.device}: ${file.path}`)
      .join('\n');
    alert(`Selected files:\n${selectedPaths}`);
  };

  return (
    <StyledTableTop>
      <input
        aria-label="Select all available files"
        type="checkbox"
        checked={selectAll}
        ref={(input) => {
          if (input) input.indeterminate = indeterminate;
        }}
        onChange={handleSelectAllChange}
        style={{ marginLeft: '-4px' }}
      />
      <SelectedText>
        {selectedRows.length > 0
          ? `Selected ${selectedRows.length}`
          : 'None Selected'}
      </SelectedText>
      <div>
        <DownloadButton
          aria-label={'Download Selected files'}
          label={'Download Selected'}
          handleOnClick={handleDownloadSelected}
          disabled={selectedRows.length === 0}
        />
      </div>
    </StyledTableTop>
  );
};
