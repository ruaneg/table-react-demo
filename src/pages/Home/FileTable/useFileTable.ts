import { useEffect, useState } from 'react';
import { FileData, Status } from './fileData';

export const isFileStatusAvailable = (status: string) =>
  status === Status.available;
const getAvailableFiles = (files: FileData[]) =>
  files.filter((file) => isFileStatusAvailable(file.status));

const useFileTable = (files: FileData[]) => {
  const [selectedRows, setSelectedRows] = useState<FileData[]>([]);
  const [selectAll, setSelectAll] = useState(false);
  const [indeterminate, setIndeterminate] = useState(false);

  useEffect(() => {
    const availableFiles = getAvailableFiles(files);
    setSelectAll(selectedRows.length === availableFiles.length);
    setIndeterminate(
      selectedRows.length > 0 && selectedRows.length < availableFiles.length
    );
  }, [selectedRows, files]);

  const handleSelectAllChange = () => {
    const availableFiles = getAvailableFiles(files);
    setSelectedRows(
      selectedRows.length === availableFiles.length ? [] : availableFiles
    );
  };

  const handleSelectChange = (file: FileData) => {
    setSelectedRows(
      selectedRows.includes(file)
        ? selectedRows.filter((f) => f !== file)
        : [...selectedRows, file]
    );
  };
  return {
    selectedRows,
    handleSelectChange,
    selectAll,
    handleSelectAllChange,
    indeterminate,
  };
};

export default useFileTable;
