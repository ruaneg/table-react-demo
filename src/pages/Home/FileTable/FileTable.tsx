import styled from "styled-components";
import { FileData, Status } from "./fileData";
import useFileTable, { isFileStatusAvailable } from "./useFileTable";
import {
  IconCell,
  StyledTable,
  StyledTableCell,
  StyledTableRow,
} from "~/components/Table/Table.styles";
import { FaCircle } from "react-icons/fa";
import { TableHeader } from "~/components/Table/TableHeader";
import { TableTop } from "./TableTop";

const StatusText = styled.span<{ status: string }>`
  color: ${(props) => (props.status === Status.available ? "green" : "red")};
  text-transform: capitalize;
`;

export const FileTable = ({ files }: { files: FileData[] }) => {
  const {
    selectedRows,
    handleSelectChange,
    selectAll,
    handleSelectAllChange,
    indeterminate,
  } = useFileTable(files);

  return (
    <div>
      <TableTop
        selectAll={selectAll}
        indeterminate={indeterminate}
        selectedRows={selectedRows}
        handleSelectAllChange={handleSelectAllChange}
      />
      <StyledTable>
        <TableHeader headers={["", "Name", "Device", "Path", "", "Status"]} />
        <tbody>
          {files.map((file) => (
            <StyledTableRow
              key={file.name}
              selected={selectedRows.includes(file)}
            >
              <StyledTableCell>
                <input
                  aria-label={`Select ${file.name}`}
                  role="checkbox"
                  type="checkbox"
                  checked={selectedRows.includes(file)}
                  onChange={() => handleSelectChange(file)}
                  disabled={!isFileStatusAvailable(file.status)}
                />
              </StyledTableCell>
              {[file.name, file.device, file.path].map((item) => (
                <StyledTableCell key={item}>{item}</StyledTableCell>
              ))}
              <IconCell>
                {isFileStatusAvailable(file.status) ? (
                  <FaCircle color="green" />
                ) : null}
              </IconCell>
              <StyledTableCell>
                <StatusText status={file.status}>{file.status}</StatusText>
              </StyledTableCell>
            </StyledTableRow>
          ))}
        </tbody>
      </StyledTable>
    </div>
  );
};
