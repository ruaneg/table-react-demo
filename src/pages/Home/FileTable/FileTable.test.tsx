import { render, screen, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { describe, test, expect } from 'vitest';
import { FileTable } from './FileTable';
import { FileData, Status } from './fileData';

describe('FileTable', () => {
  const files: FileData[] = [
    {
      name: 'smss.exe',
      device: 'Mario',
      path: '\\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe',
      status: Status.scheduled,
    },
    {
      name: 'netsh.exe',
      device: 'Luigi',
      path: '\\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe',
      status: Status.available,
    },
    {
      name: 'uxtheme.dll',
      device: 'Peach',
      path: '\\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll',
      status: Status.available,
    },
  ];

  test('renders correctly with given files', () => {
    const { container } = render(<FileTable files={files} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders the correct number of rows', () => {
    render(<FileTable files={files} />);
    const rows = screen.getAllByRole('row');
    expect(rows).toHaveLength(files.length + 1); // including the header row
  });

  test.skip('checkbox is disabled for unavailable files', () => {
    render(<FileTable files={files} />);
    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes[0]).toBeDisabled(); // smss.exe is scheduled
    expect(checkboxes[1]).not.toBeDisabled(); // netsh.exe is available
    expect(checkboxes[2]).not.toBeDisabled(); // uxtheme.dll is available
  });

  test('selects and deselects files correctly', () => {
    const { getAllByRole } = render(<FileTable files={files} />);
    const checkboxes = getAllByRole('checkbox');
    fireEvent.click(checkboxes[1]); // Select netsh.exe
    expect(checkboxes[1]).toBeChecked();

    fireEvent.click(checkboxes[1]); // Deselect netsh.exe
    expect(checkboxes[1]).not.toBeChecked();
  });

  test.skip('applies correct styles to status text', () => {
    const { getByText } = render(<FileTable files={files} />);
    const availableStatus = getByText('Available');
    const scheduledStatus = getByText('Scheduled');

    expect(availableStatus).toHaveStyleRule('color', 'green');
    expect(scheduledStatus).toHaveStyleRule('color', 'red');
  });
});
