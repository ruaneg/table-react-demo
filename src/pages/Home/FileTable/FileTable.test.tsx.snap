// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileTable > renders correctly with given files 1`] = `
.c4 {
  width: 100%;
  border-collapse: collapse;
}

.c6 {
  padding: 8px;
  border: 1px solid #f1efef;
  border-left: none;
  border-right: none;
  font-size: 1rem;
  font-weight: 500;
  text-align: left;
}

.c5 {
  background: white;
  border: 1px solid #f1efef;
}

.c5:hover {
  background: #f1f1f1;
}

.c7 {
  padding: 8px;
  padding-right: 32px;
  font-size: 0.875rem;
}

.c8 {
  padding: 8px;
  text-align: end;
  font-size: 0.875rem;
}

.c3 {
  padding: 0 8px;
  font-size: 0.875rem;
}

.c2 {
  border: 0;
  background: none;
  box-shadow: none;
  border-radius: 0px;
  font-size: 1rem;
  cursor: pointer;
}

.c2:active {
  color: dodgerblue;
}

.c2:disabled {
  pointer-events: none;
  opacity: 0.6;
}

.c0 {
  padding: 16px;
  border: 1px solid #f1efef;
  border-bottom: none;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 8rem;
  margin: 0 1rem;
}

.c9 {
  color: red;
  text-transform: capitalize;
}

.c10 {
  color: green;
  text-transform: capitalize;
}

<div>
  <div
    class="c0"
  >
    <input
      aria-label="Select all available files"
      style="margin-left: -4px;"
      type="checkbox"
    />
    <div
      class="c1"
    >
      None Selected
    </div>
    <div>
      <button
        aria-label="Download Selected files"
        class="c2"
        disabled=""
      >
        <span
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
            />
          </svg>
        </span>
        Download Selected
      </button>
    </div>
  </div>
  <table
    class="c4"
  >
    <thead>
      <tr
        class="c5"
      >
        <th
          class="c6"
        />
        <th
          class="c6"
        >
          Name
        </th>
        <th
          class="c6"
        >
          Device
        </th>
        <th
          class="c6"
        >
          Path
        </th>
        <th
          class="c6"
        />
        <th
          class="c6"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="c5"
      >
        <td
          class="c7"
        >
          <input
            aria-label="Select smss.exe"
            disabled=""
            role="checkbox"
            type="checkbox"
          />
        </td>
        <td
          class="c7"
        >
          smss.exe
        </td>
        <td
          class="c7"
        >
          Mario
        </td>
        <td
          class="c7"
        >
          \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
        </td>
        <td
          class="c8"
        />
        <td
          class="c7"
        >
          <span
            class="c9"
          >
            scheduled
          </span>
        </td>
      </tr>
      <tr
        class="c5"
      >
        <td
          class="c7"
        >
          <input
            aria-label="Select netsh.exe"
            role="checkbox"
            type="checkbox"
          />
        </td>
        <td
          class="c7"
        >
          netsh.exe
        </td>
        <td
          class="c7"
        >
          Luigi
        </td>
        <td
          class="c7"
        >
          \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe
        </td>
        <td
          class="c8"
        >
          <svg
            color="green"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: green;"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
            />
          </svg>
        </td>
        <td
          class="c7"
        >
          <span
            class="c10"
          >
            available
          </span>
        </td>
      </tr>
      <tr
        class="c5"
      >
        <td
          class="c7"
        >
          <input
            aria-label="Select uxtheme.dll"
            role="checkbox"
            type="checkbox"
          />
        </td>
        <td
          class="c7"
        >
          uxtheme.dll
        </td>
        <td
          class="c7"
        >
          Peach
        </td>
        <td
          class="c7"
        >
          \\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll
        </td>
        <td
          class="c8"
        >
          <svg
            color="green"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: green;"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
            />
          </svg>
        </td>
        <td
          class="c7"
        >
          <span
            class="c10"
          >
            available
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
